### PR TITLE
search boost python, fix #422, fix #423

### LINKED ISSUE #422: Boost::Python export on 16.04
Hi,

Following #407, I have several errors about the fact that we export a Boost::Python dependency under a made up name.

Here is the first.

On 16.04 (CMake 3.5), I can build and install eigenpy fine with Python 3. But then, when I want to configure hpp-fcl:
```
-- eigenpy searched on  FOUND. eigenpy at /tmp/3/lib/libeigenpy.so
-- boost_python-py35 searched on  FOUND. boost_python-py35 at /usr/lib/x86_64-linux-gnu/libboost_python-py35.so
CMake Error at /usr/share/cmake-3.5/Modules/FindBoost.cmake:1677 (message):
  Unable to find the requested Boost libraries.

  Boost version: 1.58.0

  Boost include path: /usr/include

  Could not find the following Boost libraries:

          boost_python35

  No Boost libraries were found.  You may need to set BOOST_LIBRARYDIR to the
  directory containing Boost libraries or BOOST_ROOT to the location of
  Boost.
Call Stack (most recent call first):
  /tmp/3/lib/cmake/eigenpy/eigenpyConfig.cmake:123 (find_package)
  CMakeLists.txt:75 (FIND_PACKAGE)
```

Where `CMakeLists.txt:75` is `FIND_PACKAGE(eigenpy 2.2 REQUIRED)`, and `/tmp/3/lib/cmake/eigenpy/eigenpyConfig.cmake:123` is `find_package(Boost COMPONENTS python35 REQUIRED)`.

This issue is [breaking build on robotpkg](http://robotpkg.openrobots.org/rbulk/robotpkg/path/py-hpp-fcl/py35-hpp-fcl-1.4.5/Ubuntu-16.04.pal-x86_64/bulk.html)

Adding the generated `find_package(Boost COMPONENTS python35 REQUIRED)` in various places, I found that the issue is raised if this line is added before that one

https://github.com/jrl-umi3218/jrl-cmakemodules/blob/fb4c22c319ec5320f9a85527eb1a4130954846f5/boost.cmake#L61

So, should we ship this `FindBoost.cmake` ? Or do we have another solution ?
Removing Boost::Python from the exports would be a workaround too.

In any case, I think that we should ensure that the ways we find Boost and its components are the same, whether we are in the main CMakeLists.txt or in a generated xxxConfig.cmake.

### LINKED ISSUE #423: Boost::Python export on Arch
Hi,

Following #407, I have several errors about the fact that we export a Boost::Python dependency under a made up name.

Here is the second, following #422

On ArchLinux (CMake 3.18), I get:
```
-- eigenpy searched on  FOUND. eigenpy at /tmp/3/lib/libeigenpy.so
-- boost_python3 searched on  FOUND. boost_python3 at /usr/lib/libboost_python3.so
CMake Error at /usr/lib64/cmake/Boost-1.72.0/BoostConfig.cmake:120 (find_package):
  Could not find a package configuration file provided by "boost_python3"
  (requested version 1.72.0) with any of the following names:

    boost_python3Config.cmake
    boost_python3-config.cmake

  Add the installation prefix of "boost_python3" to CMAKE_PREFIX_PATH or set
  "boost_python3_DIR" to a directory containing one of the above files.  If
  "boost_python3" provides a separate development package or SDK, be sure it
  has been installed.
Call Stack (most recent call first):
  /usr/lib64/cmake/Boost-1.72.0/BoostConfig.cmake:185 (boost_find_component)
  /usr/share/cmake-3.18/Modules/FindBoost.cmake:448 (find_package)
  /usr/share/cmake-3.18/Modules/CMakeFindDependencyMacro.cmake:47 (find_package)
  /tmp/3/lib/cmake/eigenpy/eigenpyConfig.cmake:128 (find_dependency)
  CMakeLists.txt:75 (FIND_PACKAGE)
```

I'll open a PR for this.

### PR DESCRIPTION
**Pull request by nim65s**
Replace #430 

It looks like this gives the right results in those cases:
```
16.04/py2       Boost_PYTHON_LIBRARY = /usr/lib/x86_64-linux-gnu/libboost_python-py27.so
16.04/py3       Boost_PYTHON_LIBRARY = /usr/lib/x86_64-linux-gnu/libboost_python-py35.so
18.04/py2       Boost_PYTHON_LIBRARY = /usr/lib/x86_64-linux-gnu/libboost_python-py27.so
18.04/py3       Boost_PYTHON_LIBRARY = /usr/lib/x86_64-linux-gnu/libboost_python-py36.so
20.04/py3       Boost_PYTHON_LIBRARY = /usr/lib/x86_64-linux-gnu/libboost_python38.so
archlinux/py2   Boost_PYTHON_LIBRARY = /usr/lib/libboost_python27.so
archlinux/py3   Boost_PYTHON_LIBRARY = /usr/lib/libboost_python38.so
buster/py2      Boost_PYTHON_LIBRARY = /usr/lib/x86_64-linux-gnu/libboost_python27.so
buster/py3      Boost_PYTHON_LIBRARY = /usr/lib/x86_64-linux-gnu/libboost_python37.so
centos7/py2     Boost_PYTHON_LIBRARY = /usr/lib64/libboost_python-mt.so
centos7/py3     Boost_PYTHON_LIBRARY = /usr/lib64/libboost_python3-mt.so
fedora28/py2    Boost_PYTHON_LIBRARY = /usr/lib64/libboost_python.so
fedora28/py3    Boost_PYTHON_LIBRARY = /usr/lib64/libboost_python3.so
fedora31/py2    Boost_PYTHON_LIBRARY = /usr/lib64/libboost_python27.so
fedora31/py3    Boost_PYTHON_LIBRARY = /usr/lib64/libboost_python37.so
stretch/py2     Boost_PYTHON_LIBRARY = /usr/lib/x86_64-linux-gnu/libboost_python-py27.so
stretch/py3     Boost_PYTHON_LIBRARY = /usr/lib/x86_64-linux-gnu/libboost_python-py35.so
```

This is still a draft, as I need to check a few other things before merging.

### CODE REVIEW

File: boost.cmake [hunk 1]
```diff
@@ -43,26 +49,82 @@ ENDFUNCTION(SEARCH_FOR_BOOST_COMPONENT boost_python_name found)
...
+    SEARCH_FOR_BOOST_COMPONENT(${BOOST_PYTHON_COMPONENT} BOOST_PYTHON_FOUND)
+    IF(BOOST_PYTHON_FOUND)
+      SET(BOOST_PYTHON_NAME ${BOOST_PYTHON_COMPONENT})
+      BREAK()
+    ENDIF(BOOST_PYTHON_FOUND)
+  ENDFOREACH(BOOST_PYTHON_COMPONENT ${BOOST_PYTHON_COMPONENT_LIST})
+
+  # If boost-python has not been found, warn the user, and look for just "python"
+  IF(NOT BOOST_PYTHON_FOUND)
+    MESSAGE(WARNING "Impossible to check boost python version. Trying with 'python'.")
```
Comment (jcarpent) [inline boost.cmake:90]: Boost.Python for the message would be better.

File: boost.cmake [hunk 2]
```diff
@@ -43,26 +49,82 @@ ENDFUNCTION(SEARCH_FOR_BOOST_COMPONENT boost_python_name found)
...
+MACRO(SEARCH_FOR_BOOST_PYTHON)
+  CMAKE_PARSE_ARGUMENTS(_BOOST_PYTHON_REQUIRED "REQUIRED" "" "" ${ARGN})
+  SET(BOOST_PYTHON_NAME "python")
+  SET(BOOST_PYTHON_REQUIRED "")
+  IF(_BOOST_PYTHON_REQUIRED)
+    SET(BOOST_PYTHON_REQUIRED REQUIRED)
+  ENDIF(_BOOST_PYTHON_REQUIRED)
+
+  IF(NOT PYTHONLIBS_FOUND)
+    MESSAGE(FATAL_ERROR "PYTHON has not been found. You should first call FindPython before calling SEARCH_FOR_BOOST_PYTHON macro.")
```
Comment (jcarpent) [inline boost.cmake:70]: Python in the message would be better too.

File: boost.cmake [hunk 3]
```diff
@@ -23,18 +23,24 @@
...
+  SET(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake/boost ${CMAKE_MODULE_PATH})
+  MESSAGE(STATUS "CMake versions older than 3.12 may warn when looking to Boost components. Custom macros are used to find it.")
+ENDIF(CMAKE_VERSION VERSION_LESS "3.12")
+
+
 #.rst:
 # .. variable:: BOOST_COMPONENTS
 #
-#  Controls the components to be detected.  
+#  Controls the components to be detected.
```
Comment (jmirabel) [inline boost.cmake:43]:
```suggestion
# This variable is *deprecated*. See :cmake:command:`SEARCH_FOR_BOOST` for more info.
#
#  Controls the components to be detected by the deprecated macro :cmake:command:`SEARCH_FOR_BOOST`. 
```

File: boost.cmake [hunk 4]
```diff
@@ -43,26 +49,82 @@ ENDFUNCTION(SEARCH_FOR_BOOST_COMPONENT boost_python_name found)
...
+  SET(Boost_PYTHON_LIBRARY ${Boost_${UPPERCOMPONENT}_LIBRARY})
+  MESSAGE(STATUS "Boost_PYTHON_LIBRARY: ${Boost_PYTHON_LIBRARY}")
+  LIST(APPEND Boost_PYTHON_LIBRARIES ${Boost_PYTHON_LIBRARY})
+  LIST(APPEND LOGGING_WATCHED_VARIABLES Boost_PYTHON_LIBRARY)
+ENDMACRO(SEARCH_FOR_BOOST_PYTHON)
+
 #
 # .. command:: SEARCH_FOR_BOOST
 #
+#  Deprecated. See :command:`SEARCH_FOR_BOOST_PYTHON`
```
Comment (jmirabel) [inline boost.cmake:112]:
```suggestion
#  Deprecated. For Boost Python, use :command:`SEARCH_FOR_BOOST_PYTHON`. For other Boost components, use::
#
#    add_project_dependency(Boost COMPONENTS ...) # or other argument to the find_package command.
#
#  or, if you don't want it to be exported::
#
#    find_package(Boost COMPONENTS ...) # or other argument to the find_package command.
```

**Comment by jmirabel** [comment]
I made some doc suggestions but did not try to generate the doc to see if there are some typos/bugs.

**Comment by nim65s** [comment]
Thanks for your suggestions. I finished my tests for #438, and everything works for me, so if no one has other suggestions, we can merge it.

**Pull request merged**